content/blog/2023-12-11-polyhedral/index.md

```diff
@@ -0,0 +1,134 @@
+
+The implementation currently performs the following analyses, each of which is described in detail below.
+
+- Detect static control parts.
+- Construct the polyhedral representation of each statement's iteration domain.
+- Construct the dependence polyhedra for a given pair of memory references.
+
+### Building Static Control Parts
+
+Control flow in the polyhedral model comprises a restricted class of loop nests whose bounds can be specified as a set of affine inequalities, such that the iteration domain forms a convex polyhedron. The class of program fragments amenable to polyhedral representation (that is, with convex polyhedral iteration domain) is said to have *static control*. A maximal region with static control is referred to as a *static control part*.
+
+The first step in our analysis is thus the identification of these program regions. We follow the basic algorithm given by Bastoul:
+
+1. Walk the IR, identifying operations with static control: loops with affine bounds and unit step, conditionals whose conditions are affine inequalities, and memory references with affine subscripts. In all of the these cases, affine functions may depend only on surrounding induction variables, constants, and parameters not mutated within the SCoP.
+2. Merge consecutive operations with static control into static control parts.
+
+The result is a list of the SCoPs within a given function. Implementing this analysis in MLIR is relatively straightforward due to the IR's explicitly nesting regions.
+
+### Building Iteration Domains
+
+Having identified regions with polyhedral domain, the next step is to construct those domains explicitly in a format that can be manipulated by integer programming tools. We use [isl](https://libisl.sourceforge.io/) to build sets from affine constraints and discharge queries such as emptiness checking.
+
+For a given operation, we begin by determining the space in which its iteration domain should be constructed. This determination depends on the SCoP in which the operation resides: loop induction variables within the SCoP correspond to dimensions of the space, while other variables correspond to parameters.
+
+To construct the iteration domain, we traverse the loop nest moving outwards from the operation, accumulating constraints imposed by loop bounds and `if` conditions, until we reach the root of the SCoP. Each constraint we encounter introduces an inequality between affine expressions. We convert these expressions from MLIR `Value`s into isl's internal representation of affine functions, construct integer sets corresponding to the inequalities, and finally compute the intersection across all constraints.
+
+In the implementation, this analysis is wrapped up in an MLIR pass that annotates each operation in the IR with an attribute encoding the operation's iteration domain. The figure from the introduction was generated from the pass' output.
```

Comment:
Do you think you could include one or two examples of what that annotation looks like, in MLIR syntax? I know it will be messy, but I think it would be helpful to get a concrete picture of what the domain consists of, in a way that is separate from the internal isl data structures.